crates/ty_server/src/session.rs

```diff
@@ -150,6 +162,45 @@ impl Session {
         self.shutdown_requested = requested;
     }
 
+    pub(crate) fn set_suspended_workspace_diagnostics_request(
+        &mut self,
+        request: SuspendedWorkspaceDiagnosticRequest,
+        client: &Client,
+    ) {
+        self.suspended_workspace_diagnostics_request =
+            request.resume_if_revision_changed(self.revision, client);
+    }
+
+    pub(crate) fn take_suspended_workspace_diagnostic_request(
+        &mut self,
+    ) -> Option<SuspendedWorkspaceDiagnosticRequest> {
+        self.suspended_workspace_diagnostics_request.take()
+    }
+
+    /// Resumes (retries) the workspace diagnostic request if there
+    /// were any changes to the [`Session`] (the revision got bumped)
+    /// since the workspace diagnostic request ran last time.
+    ///
+    /// The workspace diagnostic requests is ignored if the request
+    /// was cancelled in the meantime.
+    pub(crate) fn resume_suspended_workspace_diagnostic_request(&mut self, client: &Client) {
+        self.suspended_workspace_diagnostics_request = self
+            .suspended_workspace_diagnostics_request
+            .take()
+            .and_then(|request| {
+                if !self.request_queue.incoming().is_pending(&request.id) {
+                    // Clear out the suspended request if the request has been cancelled.
+                    return None;
+                }
+
+                request.resume_if_revision_changed(self.revision, client)
+            });
+    }
+
+    fn bump_revision(&mut self) {
```

Comment:
I think it'd be useful to document some example cases when this should be bumped.